### PR TITLE
[Do Not Merge] New GUI Test Tools

### PR DESCRIPTION
**Pull request by corranwebster**
This is a proposed new version of the GUI test assistant that is more stand-alone and backend-independent.  The biggest changes are to rely more on `pyface.timer` (which in turn implies some changes to that API) and to push a lot of the backend dependent behaviour onto the `GUI` class (in hopefully a sensible way).

The API is tested out on the `Window` class.

Tests need to be written.

### CODE REVIEW

**Comment by corranwebster** [comment]
@mdickinson could you please have a look at the proposed API (particularly for `GUITestCase` and `EventLoopHelper`).  The main changes are:

- it is a subclass of `TestCase` rather than a mixin (I could be convinced either way)
- most toolkit-dependent event loop stuff is pushed to the GUI class
- it uses `pyface.timer`s rather than working at the toolkit level for timers
- no misleading context managers

There are some additional changes that I'd be interested in making (such as renaming `event_loop_...` methods to just `loop_...`), and there is currently no replacement for the `ModalDialogTester` (and probably won't be in this PR). Also there are errors on Windows that need to be wrangled, but I'm not going to invest the effort before getting some feedback that this is a better solution than our current one.

File: pyface/testing/event_loop_helper.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        """ Suppress exit of the application when the last window is closed.
+        """
+        flag = self.gui.quit_on_last_window_close
+        self.gui.quit_on_last_window_close = False
+        try:
+            yield
+        finally:
+            self.gui.quit_on_last_window_close = flag
+
+    def event_loop(self, repeat=1, allow_user_events=True):
```
Comment (mdickinson) [inline pyface/testing/event_loop_helper.py]: I'd really like to include some advice in the docstring that if you find yourself needing to use a `repeat` of greater than `1`, you should consider whether you can use `event_loop_until_condition` instead.

File: pyface/testing/event_loop_helper.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        """ Emulate an event loop running ``repeat`` times.
+
+        Parameters
+        ----------
+        repeat : positive int
+            The number of times to call process events.  Default is 1.
+        allow_user_events : bool
+            Whether to process user-generated events.
```
Comment (mdickinson) [inline pyface/testing/event_loop_helper.py]: Perhaps add "for example keyboard and mouse events" to clarify what's meant by "user-generated events" in this context?

File: pyface/testing/event_loop_helper.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+        This runs the real event loop, rather than emulating it with
+        :meth:`GUI.process_events`.  Conditions and timeouts are tracked
+        using timers.
+
+        Parameters
+        ----------
+        condition : callable
+            A callable to determine if the stop criteria have been met. This
+            should accept no arguments.
```
Comment (mdickinson) [inline pyface/testing/event_loop_helper.py]: I think this should also be explicit about the expected return type of the callable.

File: pyface/testing/event_loop_helper.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+        timeout : float
+            Number of seconds to run the event loop in the case that the trait
```
Comment (mdickinson) [inline pyface/testing/event_loop_helper.py]: Copypasta here: "the trait change does not occur".

File: pyface/testing/event_loop_helper.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        """
+        with self.dont_quit_when_last_window_closed():
+            repeat_timer = EventTimer.timer(
+                repeat=repeat,
+                interval=0.05,
+                expire=timeout,
+            )
+            repeat_timer.on_trait_change(self._on_stop, 'active')
+
+            try:
```
Comment (mdickinson) [inline pyface/testing/event_loop_helper.py]: Is there a way to engineer a test for the corner case where the timer timeout expires before the event loop is started? (Perhaps by patching `start_event_loop` to sleep for some time first.) I'm reasonably convinced that the current logic is fine (`_on_stop` *must* be called from the main thread, so it can't possibly be called until the event loop starts), but it would be good to have a test to prevent future regressions.

What happens if the timeout is zero? Is that a use-case we want to support, or should the docstring specify that timeout should be positive?

File: pyface/testing/event_loop_helper.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                    self._on_stop, 'active', remove=True)
+                condition_timer.stop()
+
+    def event_loop_with_timeout(self, repeat=2, timeout=10):
+        """ Run the event loop for timeout seconds.
+
+        Parameters
+        ----------
+        timeout: float, optional, keyword only
+            Number of seconds to run the event loop. Default value is 10.0.
```
Comment (mdickinson) [inline pyface/testing/event_loop_helper.py]: The `repeat` parameter isn't documented.

File: pyface/testing/event_loop_helper.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+                condition_timer.on_trait_change(
+                    self._on_stop, 'active', remove=True)
+                condition_timer.stop()
+
+    def event_loop_with_timeout(self, repeat=2, timeout=10):
+        """ Run the event loop for timeout seconds.
+
+        Parameters
+        ----------
+        timeout: float, optional, keyword only
```
Comment (mdickinson) [inline pyface/testing/event_loop_helper.py]: "keyword only" doesn't seem to actually be true here. Presumably the intent is to add that extra "*" to the signature after Python 2 support has been dropped? (If so, should there be an issue open for that?)

File: pyface/testing/event_loop_helper.py (hunk 6 above)
Comment (mdickinson) [reply pyface/testing/event_loop_helper.py]: We should also document the conditions under which `ConditionTimeoutError` is raised.

I'm not really sure I understand when this method would be used, and when it should be considered that an error has occurred.

File: pyface/testing/event_loop_helper.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        :meth:`GUI.process_events`.  Conditions and timeouts are tracked
+        using timers.
+
+        Parameters
+        ----------
+        condition : callable
+            A callable to determine if the stop criteria have been met. This
+            should accept no arguments.
+
+        timeout : float
```
Comment (mdickinson) [inline pyface/testing/event_loop_helper.py]: Should this also be documented as "keyword only", to match `event_loop_with_timeout`?

File: pyface/testing/gui_test_case.py [hunk 9]
```diff
@@ -0,0 +1,443 @@
...
+        self.addCleanup(self._delete_attrs, "gui", "app", "event_loop_helper")
+        self.addCleanup(self._restore_window_activate)
+        self.addCleanup(self._restore_quit_on_last_window_close)
+        self.addCleanup(self.gui.clear_event_queue)
+        self.addCleanup(self.gui.process_events)
+        self.addCleanup(self._close_top_level_windows)
+        self.addCleanup(self._gc_collect)
+        self.addCleanup(self.event_loop_helper.event_loop, 5, False)
+
+        super(GuiTestCase, self).setUp()
```
[the remark this answers is not in the document]
Comment (mdickinson) [reply pyface/testing/gui_test_case.py]: > But that likely won't work

Okay, thinking about this, I'm not sure any more. It may work after all. The `UnittestTools` class will just be effectively invisible to the `super` machinery, because it doesn't implement `setUp` and `tearDown` at all.

Can we have a test for multiple inheritance cases that verifies that both the parent class `setUp` methods get called as expected? One way that this could potentially break is if some future version of `UnittestTools` *does* decide to implement `setUp`, so it would be useful to have a test that alerts us to the breakage.

File: pyface/testing/gui_test_case.py [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+        interfering with another.
+        """
+
+        super(GuiTestCase, self).tearDown()
+
+    def _gc_collect(self):
+        # Some top-level widgets may only be present due to cyclic garbage not
+        # having been collected; force a garbage collection before we decide to
+        # close windows. This may need several rounds.
+        for _ in range(10):
```
Comment (mdickinson) [inline pyface/testing/gui_test_case.py]: It may be worth adding an "else" clause here that complains loudly (perhaps with a `warnings.warn` warning of some kind) if we reach it; if 10 rounds of garbage collection still haven't cleaned up everything, then something's likely seriously wrong (e.g., a separate thread that's busily creating new cycles while we're trying to clean them up) and worthy of investigation.

File: pyface/testing/util.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+                sys.stdout = _old_stdout
+                sys.stderr = _old_stderr
+
+
+@contextmanager
+def _convert_none_to_null_handle(stream):
+    """ If 'stream' is None, provide a temporary handle to /dev/null. """
+
+    if stream is None:
+        out = open(os.devnull, 'w')
```
Comment (mdickinson) [inline pyface/testing/util.py]: For Python 3, I suggest adding an explicit encoding here. Seems nitpicky, but there's at least a possibility that the system encoding doesn't allow arbitrary Unicode to be written.

**Comment by mdickinson** [comment]
I left a few comments on the code, mostly (but not entirely) at nitpick level.

> * it is a subclass of `TestCase` rather than a mixin (I could be convinced either way)

Me too; there are tradeoffs either way, but overall I think I'm happier with the subclass approach. If it makes `GuiTestCase` usable with the usual `super` pattern in a multiple inheritance situation, so much the better.

Overall the API looks good to me, but I don't think I understand what `EventLoop.event_loop_with_timeout` and `GuiTestCase.event_loop_with_timeout` are for. From just the name and the signature, I'd assume that its for cases where you want to say "I've set things up so that the event loop will only run for a short finite time. Now either go and run the event loop until it stops naturally, or if it hasn't stopped after <timeout> seconds raise a timeout exception." But that doesn't seem to be what the code is doing.

Will there be unit tests for the `EventLoop` and the `GUITestCase`?

**Comment by mdickinson** [comment]
N.B. I didn't really look at the parts of this PR that aren't directly related to the `GuiTestCase`.

**Comment by corranwebster** [comment]
Yes, `event_loop_with_timeout` was problematic, and the docstring got out of sync with the final version.  The original intent was as in the current `GuiTestAssistant`: run the actual event loop at least `repeat` times with a timeout if it takes too long to repeat that many times.

I then went to the start which the current docstring suggests which was "run for `timeout` seconds" no matter what due to issues getting it to work on WxPython; but it turned out that the problem was in fact that under WxPython you can't _start_ the event loop until there is at least one window in existence; so I reverted to the original intent and used a different approach for the use case which was causing problems.

But given that we are using it in just one place, it maybe needs a re-think or removal (even the old version is not used much).  After all, without the repeat condition, it is just `event_loop_with_condition` where the condition always returns `False`.
